### PR TITLE
Instructor: feedback results: provide copy instructions for Mac OS #7350

### LINKED ISSUE #7350: Instructor: feedback results: provide copy instructions for Mac OS
In instructorFeedbackResultsPage, when using the "click here" text to view the results in a table, instructions are provided to copy and paste the table into an Excel workbook.
feedbacktable
However, Mac systems use `Command + C` and `Command + V` instead, which could be added to the instructions.

### PR DESCRIPTION
**Pull request by ncbaihly**
Fixes #7350 

Added text and HTML encoding of Mac command key symbol to instruct Mac users on copying and pasting from tables. 

mac_copy_tips

I also used GodMode to change the corresponding expected test results so that the tests correctly evaluate the new changes.

### CODE REVIEW

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag [hunk 1]
```diff
@@ -4,10 +4,10 @@
   <div class="modal-dialog modal-lg">
     <div class="modal-content">
       <div class="modal-header">
-        <span class="pull-left help-block">
-          Tips: After Selecting the table, <kbd>Ctrl + C</kbd> to COPY and <kbd>Ctrl + V</kbd> to PASTE to your Excel Workbook.
+        <span class="help-block">
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: What are the effects of removing `pull-left`?

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag [hunk 2]
```diff
@@ -4,10 +4,10 @@
...
+          Tips: After Selecting the table, <kbd>Ctrl + C</kbd> or <kbd>&#8984; + C</kbd> to COPY and
+          <kbd>Ctrl + V</kbd> or <kbd>&#8984; + V</kbd> to PASTE to your Excel Workbook.
         </span>
-
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: This is an unrelated change.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag [hunk 3]
```diff
@@ -17,11 +17,10 @@
...
     <div class="modal-content">
       <div class="modal-header">
-        <span class="help-block" style="display:inline;">
-          Tips: After selecting the table, <kbd>Ctrl + C</kbd> to COPY and
-          <kbd>Ctrl + V</kbd> to PASTE to your Excel Workbook.
+        <span class="help-block">
+          Tips: After selecting the table, <kbd>Ctrl + C</kbd> or <kbd>&#8984; + C</kbd> to COPY and
+          <kbd>Ctrl + V</kbd> or <kbd>&#8984; + V</kbd> to PASTE to your Excel Workbook.
         </span>
-        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: This is an unrelated change.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag [hunk 4]
```diff
@@ -17,11 +17,10 @@
   <div class="modal-dialog modal-lg">
     <div class="modal-content">
       <div class="modal-header">
-        <span class="help-block" style="display:inline;">
-          Tips: After selecting the table, <kbd>Ctrl + C</kbd> to COPY and
-          <kbd>Ctrl + V</kbd> to PASTE to your Excel Workbook.
+        <span class="help-block">
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: What are the effects of removing `display:inline`?

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag (hunk 1 above)
Comment (ncbaihly) [reply src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: pull-left is a bootstrap CSS class that floats the span left, I think this was done to have a better appearance with the "close" and "select table" buttons on the same line as the tip. 

Since the tip is longer it looked awkward with the buttons over on the side so far so I decided to have them on the line below. In that case, I think it's better just to have the tip text over the buttons. So I just thought it looks better. I understand though that it might not fit if you have a specific set of rules for design stuff.

(There are pictures of the original and my change in the issue comments for reference.)

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag (hunk 2 above)
Comment (ncbaihly) [reply src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: Should I revert this back to the original? Sorry this might have been a habit that resurfaced itself based on the style guide we used at a place I worked.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag (hunk 4 above)
Comment (ncbaihly) [reply src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: display:inline was being used to override the display:block that this element gets from the help-block CSS class. This was to make sure that the tip text and buttons appeared on the same row.

Since I was moving the buttons below the now longer tip text, this inline styling could be removed. Also looking at your linked HTML style guide, inline styling should be avoided when possible and with my changes inline styling is no longer necessary. 

Also, I understand that there might be some reason why copyPasteModal.tag and studentTableModal.tag are separate files that I don't know about since I'm new. But it seems to me that the tip part could be merged into one file so that future changes for this would only have to be made in one file. 

But again, I understand that there might be some reason unbeknownst to me why this might be done on purpose.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag (hunk 3 above)
Comment (ncbaihly) [reply src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: The reason I removed this is because it seemed extraneous and it serves no purpose now that the tip is displayed as a block with the buttons below it. 

This seems to have been used to ensure that there was some space between the end of the tip text and the buttons when they were on the same line.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag (hunk 4 above)
Comment (whipermr5) [reply src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: There may not be any good reason, other than the small differences between the two (e.g. the IDs are`studentTableWindow` vs `fsResultsTableWindow` and `detailsTable` vs `fsModalTable`). You are welcome to combine them into a single tag file.

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag (hunk 3 above)
Comment (whipermr5) [reply src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: Apologies; I didn't realise the change in UI (the tip now being displayed as a block). This is fine then.

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag (hunk 2 above)
Comment (whipermr5) [reply src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: Yes, unless you decide to combine the two tag files into one, in which case the two originals can be deleted.

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag (hunk 1 above)
Comment (whipermr5) [reply src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: Apologies; I only just closely reviewed the UI change by pulling your code and viewing the dialog in question. A suggestion I have is to use a small cross on the top right of the dialog instead of a close button, and a link instead of a select button. This would help to save space and results in a more streamlined appearance:
screen shot 2017-10-27 at 4 39 03 pm
You can refer to the edit view dialog for the required HTML to implement the top right cross.

While this is a lot more than we typically demand of a first time contributor, would you be willing to give it a shot? You seem to be pretty proficient in this :)

Comment (ncbaihly) [reply src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: That is a good suggestion and I agree that would be an improvement. Yes, I'd be willing to take on that task. Thanks :)

**Comment by LiHaoTan** [comment]
@ncbaihly I'm just wondering if there is a particular reason to use the character entity reference for ⌘ in the HTML? Why not use the glyph ⌘ directly?

**Comment by ncbaihly** [comment]
@LiHaoTan To be honest I just kind of did it out of habit. Given how widely unicode is supported I think I will switch to using the glyph directly instead.

**Comment by ncbaihly** [comment]
So I've run into a display issue. I've merged copyPasteModal and studentTableModal into one file. And I'm using the utf-8 character for ⌘(U+2318), however it is not displaying correctly in the browser. It shows up like the below screenshot. 
encoding-issue

I'm actively working on the issue, however any input on what might be causing this would be welcome. 

Also my project text encoding is set to utf-8 and I noticed `<meta charset="utf-8">` is used on the page.

**Comment by whipermr5** [comment]
@ncbaihly Could you push your code so I can take a look? Also, can try running it in the [new Java 8 environment](https://github.com/TEAMMATES/teammates/issues/8149).

**Comment by ncbaihly** [comment]
@whipermr5 I'll push my code up when I get the chance. Also, I tried merging the two files, and as it is right now I think maybe for this issue I'll just keep them separate. 

It was a little more than I could deal with at the moment( would also have to figure out a solution with the Javascript which would also mean trying to merge those two files into one which I think is a bit excessive ).

**Comment by whipermr5** [comment]
@ncbaihly Hmm the problem seems to lie with the new Java 8 environment rather than this PR - even the existing UTF-8 characters in `instructorHelp.jsp` aren't displaying correctly in the latest release. Investigating.

**Comment by wkurniawan07** [comment]
@whipermr5 likely it's [this line I removed when migrating](https://github.com/TEAMMATES/teammates/commit/ca27ffb83f6eca9ca7df67d292554f420f6e09c1#diff-411d1a9625fc3f5d5be34d116942f6caL11). It was giving a problem before but I could not remember why. Also, it was done way before Java 8 in GAE standard gains GA status so whatever problem I faced before may have been fixed now.

**Comment by ncbaihly** [comment]
@whipermr5 When I committed my stuff and was having issues with it, I was still on the old environment. I was going to switch to Java 8 today. 

May not be important but thought I'd mention it.

**Comment by wkurniawan07** [comment]
@whipermr5 I just tried added that line back again and there doesn't seem to be anything wrong, so it really should have been fixed from the SDK side.
My guess, though, the problem is because the page encoding setting applies to `*.jsp` files, but not `*.tag` files, which is why even on the Java 7 environment it doesn't work.

**Comment by whipermr5** [comment]
@wkurniawan07 Adding back that line leads to a non-functional app when deployed:
screen shot 2017-11-19 at 2 42 37 pm
The error in the log is the one mentioned here: https://stackoverflow.com/questions/46391970/getting-illegalstateexception-no-such-servlet-jsp-when-accessing-deployed-spri - from this link I also learned, and verified, that `<trim-directive-whitespaces>` actually _does_ work for `jsp`s (but not for `tag`s) on the old Java 7 runtime.

**Comment by wkurniawan07** [comment]
@whipermr5 right, that's the error I encountered before.
Then it seems like our only option is to specify the encoding on every single JSP/tag files? While doing that we can also add `trimDirectiveWhitespaces="true"`.

**Comment by whipermr5** [comment]
@wkurniawan07 My sentiments exactly 👍 Will do it in V6.0.1rc. @ncbaihly Meanwhile you can modify the two `tag` files to use `<%@ tag description="..." pageEncoding="UTF-8" %>` and the build should pass.

**Comment by ncbaihly** [comment]
@whipermr5 The glyphs are working fine now :) and I was about to update the tests using god mode, however I noticed you changed the case of "selecting" in copyPasteModal.tag but not in studentTableModal.tag.
ex. "Tips: After Selecting the table," --> "Tips: After selecting the table,"

Maybe it's something you missed, I was going to go ahead and make the change in studentTableModal.tag, but just thought I'd check in with you.

File: src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag [hunk 5]
```diff
@@ -4,13 +4,13 @@
...
-          Close
+        <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+            <span aria-hidden="true">&times;</span>
         </button>
+        <span class="help-block">
+          Tips: After selecting the table, <kbd>Ctrl + C</kbd> or <kbd>⌘ + C</kbd> to COPY and
+          <kbd>Ctrl + V</kbd> or <kbd>⌘ + V</kbd> to PASTE to your Excel Workbook.
+        </span>
 
         <button type="button" class="btn btn-primary" id="btn-select-element-contents">
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/course/studentTableModal.tag]: Can we use a link instead of a button, as in my original suggestion?
screen shot 2017-10-27 at 4 39 03 pm

The content is already automatically selected, so we don't even expect the user to use this option under normal circumstances. A link would suffice. Just do `<span class="btn-link" id="btn-select-element-contents">selecting</span>`

File: src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag [hunk 6]
```diff
@@ -21,12 +21,9 @@
             <span aria-hidden="true">&times;</span>
         </button>
         <span class="help-block">
-          Tips: After selecting the table, <kbd>Ctrl + C</kbd> or <kbd>⌘ + C</kbd> to COPY and
-          <kbd>Ctrl + V</kbd> or <kbd>⌘ + V</kbd> to PASTE to your Excel Workbook.
+          Tips: <span class="btn-link" id="btn-select-element-contents">Click here</span> to select the table, 
+          <kbd>Ctrl + C</kbd> or <kbd>⌘ + C</kbd> to COPY and <kbd>Ctrl + V</kbd> or <kbd>⌘ + V</kbd> to PASTE to your Excel Workbook.
```
Comment (whipermr5) [inline src/main/webapp/WEB-INF/tags/instructor/results/copyPasteModal.tag]: For consistency, add a `then` to the start of this line.

**Comment by whipermr5** [comment]
For reference, the final UI after the changes looks like this:
screen shot 2017-11-23 at 3 13 19 pm
